sw/host/opentitanlib/src/io/eeprom.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+
+use super::spi::SpiError;
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub enum DataWidth {
+    Single,
+    Dual,
+    Quad,
+    Octo,
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct Cmd {
+    data: [u8; 8],
+    cmd_len: u8,
+    cmd_width: DataWidth,
+    addr_len: u8,
+    addr_width: DataWidth,
+    dummy_cycles: u8,
+    data_width: DataWidth,
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct Mode {
+    pub cmd_width: DataWidth,
+    pub addr_len: u8,
+    pub addr_width: DataWidth,
+    pub dummy_cycles: u8,
+    pub data_width: DataWidth,
+}
```

Comment:
I'm a bit concerned about this abstraction. It seems like there are 3 independent variables captured in `Mode`:  The cmd/addr/data transfer style, the address phase size (3b/4b) and the number of dummy cycles.

The `addr_len` is a stateful property of the EEPROM for opcodes like `READ`, `ERASE` and `PAGE_PROGRAM`.  The SPI host usually commands the eeprom to go into a certain mode via the `EN4B` and `EX4B` opcodes.  

There are opcodes like `READ4B` that _always_ have a 4-byte address phase regardless of the 3b/4b mode. 
 Similarly, the number of dummy cycles is a feature of the `FAST_READ*` family of opcodes.

The opcodes which operate with alternate lane widths (dual and quad) are combinatorial with the address mode and dummy cycle selection (That is to say D_READ and Q_READ use 112 and 114 widths regardless of the address mode and dummy-cycle setup).

I think this `Mode` can be fine for informing the hyperdebug SPI master how to execute a given command, but its use in this PR as a replacement for AddressMode seems incorrect to me.  I think it's more that AddressMode and the other properties of a given opcode should go into constructing an appropriate `Mode` for that opcode.

One might imagine the (pseudo-code) for an eeprom quad read function looking something like this:
```
fn quad_read(&self, spi: &dynTarget, buffer: &mut [u8]) -> Result<()> {
  ensure!(target.supports(114), "The debugger's SPI master doesn't support 114 mode");
  ensure!(self.sfdp.support_114, "quad read not supported on this eeprom");

  let opcode = self.sfdp.param_114.opcode;
  let mode = Mode {
    cmd_width: 1,
    addr_len: self.address_mode,
    addr_width: 1
    dummy_cycles: self.sfdp.param_114.wait_states,
    data_width: 4,
  };
  for chunk in buffer.chunks_mut(...) {
    let op_addr = self.opcode_with_address(opcode, address)?;
    spi.run_transaction_with_mode(mode, &[Transfer::Write(&op_addr), Transfer::Read(chunk)])?;
    // calculate next addr based on chunk size
  }
}
```
(obviously, I'm projecting the pre-existing `run_transaction` API into the inner loop - it might look a bit different depending on how you think the API should look at that layer).

Reply:
Thanks for the explanation, my knowledge of multi-lane eeprom spi comes only from looking at what the STM32L5 is capable of doing, which necessarily is a superset of what actually makes sense.

I see that it was a mistake to replace `AddressMode` in flash.rs with this `Mode`, I will revert that part.

I do not like representing the op-code+address write and subsequent data read as separate Write() and Read() entries on a list.  When dealing with flash writes it could be inconvenient to put both the command and and data into a single [u8], and passing them separately results in ambiguity, e.g. a list of three `Write`s could either be a command without data followed by a write with data, or it could be a write with data followed by a command without data.  I find it more informative to have the `Cmd` and separate in/out-bound data buffer, as in my `eeprom::Transfer` enum.

I see that there is a relation between commands and bit-widths (as the D_READ and Q_READ you mention). I had considered creating an opcode enum in the Spi trait, and encoding information about the bit-widths.  However, I think that it could quickly get quite complicated with different protocol variants.  Rather, I would like to proceed with the caller explicitly passing both opcode as bytes as well as address/data modes.

Let me try to rework this PR.

Reply:
I agree that a fixed table of opcode properties would be sub-optimal.

Very few of the spi flash opcodes are truly standardized.  Anything having to do with reading or programming in any multi-lane mode should be checked in the SFDP table for that eeprom.  We _should_ also be consulting the SFDP for the various forms of `ERASE` opcodes.

I _believe_ that even though the multi-lane opcodes are not standardized, many of the vendors implement a compatible set of opcodes, **but**, we really should consult the SFDP.  I think the current SFDP implementation in `sfdp.rs` is reasonably complete and up-to-date, even if a bit obtuse.  I think everything we need for read and erase is encoded in the base table and the `rev_b` extensions.  For program opcodes, we might need a table of opcodes.

Reply:
Ok.  I have rolled back the changes to address_mode in flash.rs, and I have changed the types, to make it clear that dummy_cycles only makes sense in the multi-lane modes.  (It could still be that while exploring how HyperDebug and Quad-spi flash chips work, that I discover a need to revise something.)

The user of io/eeprom.rs is fully responsible for ensuring that opcodes and modes used fit the particular chip.  I have not yet tried to use SFDP data to choose the correct erasing and programming opcodes in flash.rs.  I have merely taken part of the low-level logic of flash.rs and factored into the default implementation of the new trait method, allowing the Dediprog implementation to tweak that part of the code.

I believe that the current API should allow me to implement a Dediprog backend, and also with good likelihood some Quad-spi functionality on HyperDebug.